File: ruby/day_12.rb
# frozen_string_literal: true

$dp = {}

def score(dots, blocks, dots_index, blocks_index, broken_length)
  key = [dots_index, blocks_index, broken_length]

  if $dp[key]
    return $dp[key]
  end

  if dots_index == dots.length
    if blocks_index == blocks.length && broken_length == 0
      return 1
    elsif blocks_index == blocks.length - 1 && blocks[blocks_index] == broken_length
      return 1
    else
      return 0
    end
  end

  answer = 0

  dot = dots[dots_index]

  for symbol in ['.', '#']
    if dot == symbol || dot == '?'
      if symbol == '.' && broken_length == 0
        answer += score(dots, blocks, dots_index + 1, blocks_index, 0)
      elsif symbol == '.' && broken_length > 0 && blocks_index < blocks.length && blocks[blocks_index] == broken_length
        answer += score(dots, blocks, dots_index + 1, blocks_index + 1, 0)
      elsif symbol == '#'
        answer += score(dots, blocks, dots_index + 1, blocks_index, broken_length + 1)
      end
    end
  end

  $dp[key] = answer

  answer
end

def valid?(dots, blocks)
  broken_count = 0
  seen = []
  dots.each_char do |dot|
    if dot == '.'
      seen.append(broken_count) if broken_count.positive?
      broken_count = 0
    elsif dot == '#'
      broken_count += 1
    end
  end
  seen.append(broken_count) if broken_count.positive?

  seen == blocks
end

def parse(line)
  dots, blocks = line.split
  blocks = blocks.split(',').map(&:to_i)
  [dots, blocks]
end

def pt1(filename)
  lines = File.open(filename).readlines.map(&:chomp)

  answer = 0

  lines.each do |line|
    dots, blocks = parse(line)

    $dp = {}

    answer += score(dots, blocks, 0, 0, 0)
  end

  answer
end

def pt2(filename)
  lines = File.open(filename).readlines.map(&:chomp)

  answer = 0

  lines.each do |line|
    dots, blocks = parse(line)

    dots = Array.new(5, dots).join('?')

    blocks *= 5

    $dp = {}

    answer += score(dots, blocks, 0, 0, 0)
  end

  answer
end

puts(pt1('12_example.txt'))
puts(pt1('12.txt'))

puts(pt2('12_example.txt'))
puts(pt2('12.txt'))

File: src/day_12.rs
use std::fs;

pub fn run() {
    let contents = fs::read_to_string("12.txt").unwrap();
    println!("pt1: {}", pt1(&contents));
    println!("pt2: {}", pt2(&contents));
}

fn valid(springs: &Vec<char>, pattern: &Vec<usize>) -> bool {
    let mut broken_count = 0;
    let mut pattern_index = 0;

    for char in springs {
        let char = *char;
        if char == '#' {
            if pattern_index == pattern.len() {
                return false
            }
            broken_count += 1;
        } else if broken_count > 0 { 
            if broken_count != pattern[pattern_index] {
                return false
            }

            broken_count = 0;
            pattern_index += 1;
        }
    }

    if broken_count > 0 && pattern_index < pattern.len() {
        if broken_count != pattern[pattern_index] {
            return false
        }

        broken_count = 0;
        pattern_index += 1;
    }

    broken_count == 0 && pattern_index == pattern.len()
}

fn score(springs: &Vec<char>, pattern: &Vec<usize>, index: usize) -> usize {
    if index == springs.len() {
        if valid(springs, pattern) {
            return 1
        } else {
            return 0
        }
    } else {
        let char = springs[index];

        if char == '?' {
            let mut answer = 0;

            let mut springs1 = springs.clone();
            springs1[index] = '#';

            answer += score(&mut springs1, &pattern, index + 1);

            let mut springs2 = springs.clone();
            springs2[index] = '.';

            answer += score(&mut springs2, &pattern, index + 1);

            return answer
        } else {
            return score(springs, &pattern, index + 1);
        }
    }
}

fn pt1(contents: &str) -> usize {
    let mut answer = 0;

    for line in contents.lines() {
        let (springs, pattern) = line.split_once(' ').unwrap();
        let mut springs: Vec<_> = springs.chars().collect();
        let pattern = pattern.split(',').map(|s| s.parse::<usize>().unwrap()).collect();
        let s = score(&mut springs, &pattern, 0);
        // println!("{} = {}", line, s);
        answer += s;
    }

    answer
}

fn pt2(contents: &str) -> usize {
    0
}

#[cfg(test)]
mod day_12_pt_1_tests {
    use super::*;

    #[test]
    fn test_example_input() {
        let example_input = "???.### 1,1,3
.??..??...?##. 1,1,3
?#?#?#?#?#?#?#? 1,3,1,6
????.#...#... 4,1,1
????.######..#####. 1,6,5
?###???????? 3,2,1";

        assert_eq!(21, pt1(example_input));
    }

    #[test]
    fn test_score_for_last_line_of_example_input() {
        let springs = vec!['?', '#', '#', '#', '?', '?', '?', '?', '?', '?', '?', '?'];
        let pattern = vec![3, 2, 1];

        assert_eq!(10, score(&springs, &pattern, 0));
    }

    #[test]
    fn test_valid() {
        let pattern = vec![3, 2, 1];

        let invalid_permutations = vec![
            vec!['.', '#', '#', '#', '.', '#', '#', '.', '#', '.', '#', '#'],
            vec!['.', '#', '#', '#', '.', '#', '#', '.', '#', '.', '.', '#'],
            vec!['.', '#', '#', '#', '.', '#', '#', '.', '.', '#', '.', '#'],
            vec!['.', '#', '#', '#', '.', '.', '#', '#', '.', '#', '.', '#'],
            vec!['.', '#', '#', '#', '.', '.', '.', '.', '.', '.', '#', '#'],
        ];

        let valid_permutations = vec![
            vec!['.', '#', '#', '#', '.', '#', '#', '.', '#', '.', '.', '.'],
            vec!['.', '#', '#', '#', '.', '#', '#', '.', '.', '#', '.', '.'],
            vec!['.', '#', '#', '#', '.', '#', '#', '.', '.', '.', '#', '.'],
            vec!['.', '#', '#', '#', '.', '#', '#', '.', '.', '.', '.', '#'],
            vec!['.', '#', '#', '#', '.', '.', '#', '#', '.', '#', '.', '.'],
            vec!['.', '#', '#', '#', '.', '.', '#', '#', '.', '.', '#', '.'],
            vec!['.', '#', '#', '#', '.', '.', '#', '#', '.', '.', '.', '#'],
            vec!['.', '#', '#', '#', '.', '.', '.', '#', '#', '.', '#', '.'],
            vec!['.', '#', '#', '#', '.', '.', '.', '#', '#', '.', '.', '#'],
            vec!['.', '#', '#', '#', '.', '.', '.', '.', '#', '#', '.', '#'],

        ];

        for permutation in invalid_permutations {
            assert!(!valid(&permutation, &pattern));
        }

        for permutation in valid_permutations {
            assert!(valid(&permutation, &pattern));
        }
    }
}

